pkg/concurrency/stopper_test.go

```diff
@@ -0,0 +1,128 @@
+package concurrency
+
+import (
+	stdSync "sync"
+	"testing"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/suite"
+)
+
+type StopperTestSuite struct {
+	suite.Suite
+}
+
+func TestStopper(t *testing.T) {
+	suite.Run(t, &StopperTestSuite{})
+}
+
+// TestCommonCase demonstrates how to use Stopper for implementing a gracefully stoppable goroutine.
+func (s *StopperTestSuite) TestCommonCase() {
+	stopper := NewStopper()
+
+	countdown := stdSync.WaitGroup{}
+	countdown.Add(6)
+
+	// This is how to implement a gracefully stoppable goroutine.
+	go func() {
+		// The goroutine must report that it shut down at the end.
+		defer stopper.Flow().ReportStopped()
+		// Here ticker simulates some async source of data that the goroutine handles.
+		ticker := time.NewTicker(50 * time.Millisecond)
```

Comment:
This might be a bit pedantic, but why not use 50 microseconds here?

Since this is a unit test with nothing else happening other than waiting, the test will needlessly wait `300ms`.

Reply:
Shorter timeout would mean more chance of flake if something in threading or in runtime takes a bit longer.
I don't mind to decrease but I don't know how to reason about which value is sufficiently big to keep the test stable.

I tried running 300 times with 1ms delay (`go test -count=300 -v -run ^TestStopper$ ./pkg/concurrency`) and the test flaked.
10 ms - no flake.
5 ms - no flake.
3 ms - no flake.
2 ms - no flake.
1 ms - got flakes.

Therefore, 50 microseconds is definitely too short.
I also don't feel very comfortable about 10ms.

I suggest to decrease the countdown from 6 to 3 and delay from 50 to 20ms.

Reply:
Well. `sync` package from standard Go library is banned by our linters therefore I had to change the implementation so that it doesn't rely on timing.

Reply:
> Well. sync package from standard Go library is banned by our linters therefore I had to change the implementation so that it doesn't rely on timing.

We have out own `pkg/sync` which wraps the standard sync functions. You should use that one instead

Reply:
Please check the recent code. There is no use of standard `sync`.

Reply:
Looks good, thanks.